### PR TITLE
UpdateEngine v0.3.6

### PR DESCRIPTION
**Pull request by ryanelian**
> I changed all ex.messages to output the stack trace too. Maybe it's an internal issue and the string is just being passed from your DLL? (git blame am i rite)

[AIDA blames my engine for the obscure exception message](https://github.com/Arks-Layer/PSO2Tweaker/pull/32#issuecomment-238030913), hence this PR. Now you can try to play around with engine exceptions :D

- Added `UpdateManager.Crash` that simply throws an exception. Might be useful if you're trying to develop a working exception handler for the engine that doesn't say `Microsoft.VisualBasic.ApplicationServices.UnhandledExceptionEventArgs`.

- "FileX deleted." message when restoring backup files goes into `IRenderer.AppendLog` instead of `WriteLine` now.

----

EDIT: Also, when testing this, try to crash this using an asynchronous event handler. I just remembered an old lecture about how `async void` methods don't handle Exceptions the same way as a synchronous method.

```c#
private async void PlayButton_Click(object sender, RoutedEventArgs e)
{
    await Task.Run(() =>
    {
        Updater.Crash();
    });
}
```

Excellion handles this with `App.xaml.cs` attaching `DispatcherUnhandledException` on constructor into an unhandled exception handler, then marking `DispatcherUnhandledExceptionEventArgs.Handled` to TRUE.

```
---------------------------
Unhandled Exception
---------------------------
An unhandled exception just occurred: Jesus bless your soul. 
   at ArksLayer.Tweaker.UpdateEngine.UpdateManager.Crash()
   at Excellion.MainWindow.<>c__DisplayClass5_0.<PlayButton_Click>b__0() in D:\VS\Excellion\Excellion\MainWindow.xaml.cs:line 42
   at System.Threading.Tasks.Task.InnerInvoke()
   at System.Threading.Tasks.Task.Execute()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.GetResult()
   at Excellion.MainWindow.<PlayButton_Click>d__5.MoveNext() in D:\VS\Excellion\Excellion\MainWindow.xaml.cs:line 40
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.AsyncMethodBuilderCore.<>c.<ThrowAsync>b__6_0(Object state)
   at System.Windows.Threading.ExceptionWrapper.InternalRealCall(Delegate callback, Object args, Int32 numArgs)
   at System.Windows.Threading.ExceptionWrapper.TryCatchWhen(Object source, Delegate callback, Object args, Int32 numArgs, Delegate catchHandler)
---------------------------
```

### CODE REVIEW

**Comment by TheCrimsonDevil** [comment]
I wasn't blaming you lol just throwing around ideas. But good stuff!

**Comment by ryanelian** [comment]
MRW when I went into forum and everyone is posting `Microsoft.VisualBasic.ApplicationServices.UnhandledExceptionEventArgs`.

fire

**Comment by ryanelian** [comment]
I think I figured out why you couldn't update GameGuard stuffs and `gameversion.ver` previously while I can.

It's the update permission issue as well. Blacklisting those files aren't the solution (and honestly I don't like blacklisting files).

**Comment by acffordyce973** [comment]
@ryanelian if people are manually installing the English patch from the psumods site (because it takes @TheCrimsonDevil forever to update to the latest) then will that create any issues come next patch? Since obviously there's no backup being created.

**Comment by ryanelian** [comment]
If the game can run with the previous English Patched files then no,
otherwise yes.

Let's say file game client hash is AAAA.
English patched game client will become AABB.
However in the next update, game becomes ACCA officially.
Since there are no backups, game now becomes ACCB, since only the delta
being downloaded

If the game can run with ACCB then no it won't cause issue. Otherwise, boom.

Wouldn't the solution is to allow PSUMod people to upload the patches
somewhere where the Tweaker can grab, instead of requiring the ARKS Layer
to manually maintain it?

On Aug 10, 2016 3:30 PM, "acffordyce973" <notifications@github.com> wrote:
>
> @ryanelian if people are manually installing the English patch from the
psumods site (because it takes @TheCrimsonDevil forever to update to the
latest) then will that create any issues come next patch? Since obviously
there's no backup being created.

**Comment by acffordyce973** [comment]
> Wouldn't the solution is to allow PSUMod people to upload the patches
somewhere where the Tweaker can grab

You haven't been around in this community for long have you? Long story short is that they won't allow it. AIDA has to manually grab it and host it herself.

**Comment by ryanelian** [comment]
I've been around since episode 1 but I never stepped foot into Tweaker
development until recently.

**Comment by acffordyce973** [comment]
http://www.pso-world.com/forums/showthread.php?236269-Tweaker-quantum-update-deletes-half-of-the-game

> so, tweaker new patch method finds 12k of "legacy files" and plain deletes them.

This one is clearly on you.

**Comment by ryanelian** [comment]
While I symphatized for the poor sod having to redownload half of the game,
I stand by my engine's correctness and robustness.

Here's probably what happened:

1. Dude actually has 12k legacy files and the engine safely deletes them.
2. Engine reliably detects the corrupted or missing files by full rehash.
3. Engine discovered that half of the game files actually require
redownloading. It then queues the download tasks for those files.
4. Those download tasks set the file sizes to 0KB initially then streams
the binary to it as the files get downloaded.
5. User was surprised by the sheer amount of download and checked game
folder only to discover that it shrank by half.
6. User blames legacy file deletion for this failure.

**Comment by acffordyce973** [comment]
In all honesty, if you're going to rewrite the updating method you shouldn't do it with a half-assed attempt. You wrote your method then expect AIDA to integrate it into the Tweaker. If you had done everything yourself then you'd either see how broken your method is or how broken the current implementation of it in the Tweaker is.

**Comment by ryanelian** [comment]
Like I said, I stand by my engine's correctness and robustness.

It's been heavily tested and before I made the initial pull request and
there is nothing wrong with it.

It's not broken and it's been written professionally and I'm not letting
anyone else say otherwise.

**Comment by acffordyce973** [comment]
Yet the amount of people with issues is rising and rising. Wow, please get off your high horse and admit your method is shit.

**Comment by acffordyce973** [comment]
Testing something yourself is not robust testing.

**Comment by acffordyce973** [comment]
I suppose you're that retarded that I have to give you proof:
https://www.reddit.com/r/PSO2/comments/4x13aq/am_i_the_only_one_that_thinks_removing_legacy/
https://www.reddit.com/r/PSO2/comments/4wztea/small_psa_tweaker_may_download_more_than_it_should/
http://www.bumped.org/psublog/pso2-jp-maintenance-and-patch-8102016/#comments
http://www.pso-world.com/forums/showthread.php?207248-PSO2-Tweaker-Various-enhancements/page712
http://www.pso-world.com/forums/showthread.php?236269-Tweaker-quantum-update-deletes-half-of-the-game

Not to mention the people from PSO2Proxy who are raging in our IRC channel.

**Comment by Kyle873** [comment]
How exactly was the update system tested?

Dealing with the amount of people having problems or issues with the current tweaker version is absolutely staggering right now.

In my opinion, this entire system should've been put into an opt-in tweaker version for a good couple weeks and maintenance sessions before it was released into the wild.

**Comment by ryanelian** [comment]
> Testing something yourself is not robust testing.

> How exactly was the update system tested?

I never said it's being tested by myself. It's been tested along with the Tweaker using the old update method for a while internally.

Excellion was started months after the Tweaker rolls in the new update method. It's been used since then to update my game by this process:

- Perform whole update method using Excellion.
- Launch Tweaker, ignore new update popup because Excellion didn't update game version due to my laziness.
- Instead, use Tweaker old update method to **VERIFY** the update. Old update method is the golden standard for updating the game due to not breaking my client for a long time.

EVERY time I does that, for months, if not more than a year, the result is always the same: **NO ADDITIONAL FILES GET DOWNLOADED**.

Not awhile ago, Tweaker gets flagged by defender and I saw the issue where the update method is not working for most people, so I decided to refactor `Excellion.UpdateEngine` into `ArksLayer.Tweaker.UpdateEngine` because I care about the community.

**Comment by TheCrimsonDevil** [comment]
This is fine.

@ryanelian It should have been tested with a multitude of people before being released to the wild. That's on me, though. However, your process may not be fullproof and perfect. An important aspect of making things like this is accepting that **things may go wrong**. I appreciate your work, and your commits, I do, but nothing is 100% foolproof. People will always find a way to break what is not broken, as you can see here. The issue might be your code, it might be VB's shit implementation of things, it might be permission issues, but we can't dismiss anything readily out of hand without crowdtesting.

I'll simply ask that @ryanelian re-codes a flag for using the legacy file method, which can be optionally (forcefully, for now) turned off until we can verify it's working correctly.

@acffordyce973 I understand you are frustrated, and rightfully so - You (and Kyle) have to deal with the brunt of "the tweaker isn't working please fix this" shit in the IRC channels. I get it. You have every right to be angry, but comments like "because it takes @TheCrimsonDevil forever to update to the latest" and "I suppose you're that retarded that I have to give you proof", while sometimes true (in the first case), only cause heated arguments and are typically the cause of people rage-quitting and projects being terminated.

You are not **wrong**, but please try to word things better.

@Kyle873 I agree - We'll have to approach this with a testing pool, perhaps an opt-in while keeping the other version available. The main issue is that all the people this is happening to are people who I don't have constant contact with - None of **us** are having issues, but it seems like 50% of the community is. It would be one thing if I could find out what's going on and reproduce it so I could work on it, but without any information on why it's happening (seriously, fuck you VB), it's hard as hell to work on.

That being said, I've reverted the Tweaker to the pre-QUANTUM SYSTEM version. I'll mess with this errorcatching code and see if we can get it to tell us something more than "I'm sitting here playing with my e-dick, have fun with mscorlib".

Also, I'm reading something about the QUANTUM SYSTEM breaking the item patch? What is that all about?

**Comment by ryanelian** [comment]
> However, your process may not be fullproof and perfect. An important aspect of making things like this is accepting that things may go wrong. I appreciate your work, and your commits, I do, but nothing is 100% foolproof. 

I never said my engine is perfect and foolproof. I merely pointed out that I stand by its reliability in times of game update.

Like I said before, people are **more than welcome** to criticize every lines of my code.

I agree with the fact that nothing is 100% foolproof. By putting the code into the Tweaker code base I intend to improve it by public exposure.

> I'll simply ask that @ryanelian re-codes a flag for using the legacy file method, which can be optionally (forcefully, for now) turned off until we can verify it's working correctly.

Sure, I'll do this later :+1: 

> Also, I'm reading something about the QUANTUM SYSTEM breaking the item patch? What is that all about?

Top kek.

**Comment by awksie** [comment]
I know this probably isn't helpful but QUANTUM-SYSTEM worked flawlessly for me and I'm in-game on Tweaker 3.1.x.x with no issues or 1013 or anything.

I don't use English patch tho. (*'▽') おつかれさまでした！

**Comment by flyergo** [comment]
Oha I wasn't aware this escalated here.

Just a few words: Refactoring/Rewriting always breaks stuff you least expect to break. The only thing you can do about that is to have a lot of people test it. Even people not being developers, who can't fix minor problems like Permission issues or wrong framework version installed, etc. 

A lot of people put the Tweaker.exe in the weirdest places in past, like C:\Windows, C:\Users, Downloads or even the Desktop, which shows a lot of them aren't very experienced with computers. And then things like User Account Control (UAC), Anti-Virus programs or other external things break stuff a developer wouldn't even think of.

Also: As far as I know there hasn't been a single PSO2 update, that didn't break things on someones computer for people who are using Tweaker. :rocket: 
Going at each others throats or blaming each other isn't helpful. I see the PSO2 Tweaker as highly experimental and an inofficial program for an official game. You are always free to use the Vanila Launcher for updates, which is recommend. And just as the Tweaker can break things, the english patches, item patch, PSO2Proxy etc. can break things, too.

If you want to play the game world-first after an update using Tweaker with a new PSO2 version than that's your own risk, to be honest. Nobody can ensure that a tool works with a black-box, that constantly changes itself.

I tested the Quantum System and it worked for me in my case.

Being a software-developer myself I can only say: The only thing we could maybe do to prevent this in the future, is to add a special PSO2 Tweaker-Beta version for testing, put a link to it the in right window and have a 'Report Bug' link to github/newissue somewhere so people can test and contribute to it before AIDA or anyone releases it.

TL;DR: Calm down, do not blame and drink :tea: , :coffee: or :beer: .

**Comment by TheCrimsonDevil** [comment]
Tweaker updates be like:

lol

**Comment by belldandu** [comment]
Ahaha... man. I agree with aida on wording things better as you guys have
lost enough people already in the past.

Arguements like these don't look good for people like me who would like to
get deeper in the cesspool of what is arks layer development.

Yes poor decisions were made on the update process, but that doesn't mean
you should berate a person who is contributing to the project all because
changing a few lines of code broke something... changing any bits of code
is bound to break something especially in an application of this
popularity. Instead of wasting time attacking the user who made the changes
with useless complaints you should be trying to help figure out what
exactly is causing the issues so they can be fixed. Dwelling on the past
should not be your first option here.

**Comment by acffordyce973** [comment]
@belldandu maybe you'd like to help the community by getting spammed by tons of people because they can't update and where your only option is to say "download this old version"? I stand by everything I said. While I'm trying to help people with issues @ryanelian is sitting here saying his method works yet not doing anything to prove it. If he was so sure it works he should be the one out helping people.

**Comment by TheCrimsonDevil** [comment]
Don't make me moderate the PR's, please.
